torch/csrc/jit/symbolic_variable.h

```diff
@@ -32,7 +32,18 @@ struct SymbolicVariable {
       if(g == nullptr) {
         g = inputs.at(0).value()->owningGraph();
       }
-      Node * n = g->insertNode(g->create(kind, num_outputs));
+      Node* n = g->insertNode(g->create(kind, num_outputs));
+      size_t max_depth = 0;
+      ScopePtr s;
+      for(auto n : inputs) {
+        size_t d = n.value()->node()->scope()->getDepth();
+        if(d > max_depth) {
```

Comment:
This seems like a pretty arbitrary choice for how the scope gets chosen. In all other cases the scope is set via a resource guard similar to the insert point. Why would it be different here? I am worried about using this heuristic here because if we ever change how a node gets created, we might arbitrarily break some hard-to-test visualization thing.

Reply:
@lanpa can comment on this too, but I think we don't have access to the actual insert point here. We might be able to improve this in the future but I can't think of a better option right now.

Reply:
You can get the current insert point as `g->insertPoint()`, but I do not think that is what you are saying. Can you clarify?

Reply:
As for resource guard, I have tried and confirmed that `Node->setScope` can be replaced by `WithCurrentScope`.

I am guessing the design merely from the code, the following might be wrong.

I think that the scope information for each node is correct before the graph is optimized (by observation).
After `_optimize_graph()` is called, some nodes are merged/canonicalized to new node but the scope is not appended to it. This can be verified by print(graph) in `torch/onnx/utils.py`.

The original (heuristic) fix is based on the assumption that the scope of the new node can be derived from the scope of its input nodes: The node with the deepest scope will be chosen, and that scope is assigned to the new node. I think things might go wrong if some input node has a very deep scope name while serving as other node’s input. Although among the models I tested this does not happen. 

To remove the heuristic, we should fix the empty scope problem from the beginning and make an assertion right after a node is created. If all nodes to be merged have the same scope, the scope for the new node is decided. (What if those input nodes have different scope? Is that possible? If it's not, we should add some assertion for that.)
I also noticed that there is no scope information for Graph and Block after `_trace()` is called. If we kept those information, the correct scope might be determined if the above situation happens.


@orionr @zdevito

Reply:
I'm going to land as-is and @lanpa will continue with additional pull requests to improve this. Just FYI.